### PR TITLE
Hordes no longer form from the inside of buildings.

### PR DESCRIPTION
**Pull request by CIB**
This makes it so that monsters no longer join a horde if they're on a tile that's marked as "inside" when despawning. Bit of a quick and dirty fix to the problem of hordes making certain special locations like schools and hospitals a bit too easy by despawning all the zeds into a horde and moving off when the location has been loaded into a reality bubble once. Not as good as actual pathing for hordes, but still probably better than what we currently have.

Technical details:
- Split part of game::shift_monsters off into a new function game::despawn_monsters
- Call game::despawn_monsters before map shifting happens
- Use information on whether terrain from which the monster despawned in order to determine whether the monster should join a roaming horde
- Horde joining now happens only on despawn from map, not on horde movement tick
- Created separate function game::creature_can_roam to determine whether monster should join horde

### CODE REVIEW

**Comment by Coolthulhu** [comment]
How do you handle hordes passing over zombies and integrating them? Did you disable it altogether?

And some organization problems:
* `game` is already bloated. The functionality you add to it could easily be extracted into a namespace outside `game.cpp` and in some new file (`hordes.cpp` maybe?), since it doesn't use private methods or fields
* Style, spacing etc.

File: src/game.cpp [hunk 1]
```diff
@@ -13166,39 +13171,86 @@ void game::update_stair_monsters()
...
-            // No need to shift z coords, they are absolute
-            continue;
         }
-        // Either a vertical shift or the critter is now outside of the reality bubble,
-        // anyway: it must be saved and removed.
-        despawn_monster( i );
+    }
+}
+
+void game::shift_monsters( const int shiftx, const int shifty, const int shiftz )
```
Comment (BrettDong) [inline src/game.cpp]: Compile warning: unused parameter `shiftz`

Comment (BrettDong) [inline src/game.cpp]: Compile warning: unused parameter `shiftz`

**Comment by CIB** [comment]
> How do you handle hordes passing over zombies and integrating them? Did you disable it altogether?

That was not a feature before, nor is it now. Previously, on a regular interval, all zombies on the overmap were turned into roaming groups. Now this only happens when despawning them. The only difference should be that you can no longer turn wander spawns on in an existing world and expect all monsters in your save to instantly convert into roaming groups. Instead, only monsters that were loaded into the reality bubble at some point will gradually turn into roaming groups (newly generated overmaps will still place half of its zombies as roamers). It was never an intended feature to begin with, so I don't see this as an issue.

> game is already bloated. The functionality you add to it could easily be extracted into a namespace outside game.cpp and in some new file (hordes.cpp maybe?), since it doesn't use private methods or fields

Mostly, this is a rewrite of game::shift_monsters. I just happened to split it into several functions. I would consider moving game::shift_monsters as a refactor that should be implemented and reviewed in a separate PR. I can do that if you want, but not in this PR.

> Style, spacing etc.

The style was already inconsistent between the functions I touched. Should I use the style of the function I modify, or should I use the project's style conventions and break style with the function I modify? And no, I'm not going to restyle parts of the function that I didn't modify. Again, that'd be a separate PR.

**Comment by Coolthulhu** [comment]
> The style was already inconsistent between the functions I touched. Should I use the style of the function I modify, or should I use the project's style conventions and break style with the function I modify? And no, I'm not going to restyle parts of the function that I didn't modify.

As if astyle.
No need to style parts you didn't modify (that would increase the line count a lot), but cutting and pasting a block from one function to another - and especially from one file to another - counts as modifying the block.

Proper style in non-astyled files is not a strict requirement, but I'll count it a minor issue if there are other problems (edge cases, performance issues etc.).

**Comment by kevingranade** [comment]
> That was not a feature before, nor is it now. Previously, on a regular interval, all zombies on the overmap were turned into roaming groups. Now this only happens when despawning them. The only difference should be that you can no longer turn wander spawns on in an existing world and expect all monsters in your save to instantly convert into roaming groups. Instead, only monsters that were loaded into the reality bubble at some point will gradually turn into roaming groups (newly generated overmaps will still place half of its zombies as roamers). It was never an intended feature to begin with, so I don't see this as an issue.

This was intended behavior and needs to stay, otherwise you cannot convert from non-wander-spawns to wander spawns, ever.  Only converting after encountering them utterly defeats the purpose of wander spawns.

**Comment by CIB** [comment]
Hm.. I'll have to close the PR for the time being then, because I don't know of a way to get that to work with the approach I'm using here.

**Pull request closed without merging**